[Le Nguyen Quang Dang Khoa] ip

khoahre123 opened the pull request:
# DUKE+
> A good plan today is better than a perfect plan tomorrow - Proverb ([source](https://thedigitalprojectmanager.com/project-management-quotes-inspiration/))

Managing your tasks and planning has always been a difficult task, especially when you need to track too many tasks at the same time. Worry no more, with ***DUKE+***, planning is never been easier. ***DUKE+*** comes with
* Command-line interface
* ~~Easy~~ **EXTREMELY EASY** to remember syntax 
* Fast-pace, real-time update

And most important, it is **FREE** to use 😉 

All you need to do is 
1. Download it 
2. Double-click to open the program 
3. Add your task
4. Sit back and enjoy the peacefulness 😆 

Current and upcoming features:
- [X] Managing task
- [ ] Sort the task accordingly 
- [ ] Remind the user about upcoming deadlines

Also, if you are interested in the project, you can clone it from [here](https://github.com/khoahre123/ip.git) and run the `main` method
```java
public class Duke {
    public static void main(String[] args) throws IOException {
        new Duke(PATH, DIRECTORY).run();
    }
}
 ```

Commit:
Add Gradle support

khoahre123 committed:
Level 1: Greet, Echo, Exit

khoahre123 committed:
Level 2. Add, List

khoahre123 committed:
Level 3. Mark as Done

khoahre123 committed:
Add support for tracking types of tasks

khoahre123 committed:
Update Automated Text UI Testing input and expected output

khoahre123 committed:
Create a new class for handling command error

khoahre123 committed:
Add delete function to the bot

khoahre123 committed:
Replace some common string by enum

khoahre123 committed:
Create local database file for task list

khoahre123 committed:
Implement function to help Duke understand dates and times

khoahre123 committed:
Merge branch 'branch-Level-7'

khoahre123 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

khoahre123 committed:
Refractor the code to smaller classes

khoahre123 committed:
Organize file into suitable package

khoahre123 committed:
Add unit testing for Task and TaskList classes

khoahre123 committed:
Write JavaDoc for all the classes and methods

khoahre123 committed:
Fix the JavaDoc

khoahre123 committed:
Fix coding style for all files

khoahre123 committed:
Add find keyword function for the chat bot

khoahre123 committed:
Fix JavaDoc

khoahre123 committed:
Merge branch 'branch-A-JavaDoc'

khoahre123 committed:
Merge branch 'branch-A-CodingStandard'

khoahre123 committed:
Merge branch 'branch-Level-9'

khoahre123 committed:
Add JavaDoc for new function

nnmoq commented on src/main/java/duke/util/BotStoring.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Reads and writes input history into a database file
+ */
+public class BotStoring implements Storing {
```
```suggestion
public class BotStorage implements Storage {
```

khoahre123 replied:
Thank you for your suggestion

nnmoq commented on src/main/java/duke/util/BotStoring.java:
```diff
@@ -0,0 +1,115 @@
+ */
+public class BotStoring implements Storing {
+
+    private final String path;
```
```suggestion
    private final String PATH;
```

khoahre123 replied:
Thank you but I think this need not to be capitalize, as it is not a static variable

nnmoq commented on src/main/java/duke/util/BotStoring.java:
```diff
@@ -0,0 +1,115 @@
+public class BotStoring implements Storing {
+
+    private final String path;
+    private final String directory;
```
```suggestion
    private final String DIRECTORY;
```

khoahre123 replied:
Thank you but I think this need not to be capitalize, as it is not a static variable

nnmoq commented on src/main/java/duke/util/BotStoring.java:
```diff
@@ -0,0 +1,115 @@
+
+    private final String path;
+    private final String directory;
+    private final File database;
```
```suggestion
    private final File DATABASE;
```

khoahre123 replied:
Thank you but I think this need not to be capitalize, as it is not a static variable

nnmoq commented on src/main/java/duke/util/NumericCheck.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Checks whether the input can be parsed to number
+ */
+public class NumericCheck {
```
```suggestion
public class NumericChecker {
```

khoahre123 replied:
Thank you for your suggestion

nnmoq commented on src/main/java/duke/util/BotStoring.java:
```diff
@@ -0,0 +1,115 @@
+     * @return True if the database file exists and false otherwise
+     * @throws IOException If an I/O error occur
+     */
+    public boolean isDatabaseExists() throws IOException {
```
```suggestion
    public boolean doesDatabaseExist() throws IOException {
```

khoahre123 replied:
Thank you for your suggestion

nnmoq commented on src/main/java/duke/util/Storing.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Interface represent object that can store things
+ */
+public interface Storing {
```
```suggestion
public interface Storage {
```

khoahre123 replied:
Thanks for pointing this out 👍

nnmoq commented on src/main/java/duke/util/Storing.java:
```diff
@@ -0,0 +1,59 @@
+     * @return True if the database file exists and false otherwise
+     * @throws IOException If an I/O error occur
+     */
+    boolean isDatabaseExists() throws IOException;
```
```suggestion
    boolean doesDatabaseExist() throws IOException;
```

khoahre123 replied:
Thank you for your suggestion

nnmoq reviewed (commented):
Looks pretty good, made some coding style suggestions.

khoahre123 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

khoahre123 committed:
Add gradle and checkstyle file

khoahre123 committed:
Checkstyle and change some inappropriate places

khoahre123 committed:
Update CLI interface to GUI

khoahre123 committed:
Change constructer to varags

khoahre123 committed:
Fix javadocs of the customize exception

khoahre123 committed:
Merge branch 'branch-Level-10'

albertsutz commented on data/Duke.txt:
```diff
@@ -0,0 +1,3 @@
+| D | 1 | abc | at 2/3/2002 |
```
Hi, I'm not sure whether this is a default database file or not. Perhaps you could check as I do not think that you should put the initial database file as it should be a clean slate for new users?

albertsutz commented on src/main/java/angela/Angela.java:
```diff
@@ -0,0 +1,72 @@
+        ui = null;
+        dateTable = null;
+    }
+
```
May I know the purpose of this constructor? I am not quite sure.
I am referring to `public Angela(String... args)`

albertsutz commented on src/main/java/angela/Angela.java:
```diff
@@ -0,0 +1,72 @@
+
+    /**
+     * Initialize the Duke bot
+     * @param path Relative path where the database is located
```
I think you should add a blank space between line 33 and 34

albertsutz commented on src/main/java/angela/Angela.java:
```diff
@@ -0,0 +1,72 @@
+     * @throws IOException If an I/O exception occur
+     */
+    public Angela(String path, String directory) throws IOException {
+        TaskList taskList1 = new TaskList();
```
Perhaps you can change the name of `tasklist1` to increase readability with other names such as `dummyTasklist` or `tempTasklist`?

albertsutz commented on src/main/java/angela/Angela.java:
```diff
@@ -0,0 +1,72 @@
+        return temp;
+    }
+
+    public String getResponse(String input) {
```
Don't forget to put javaDocs! This is just me being a bit picky, but it would make your code more complete.

albertsutz commented on src/main/java/angela/command/DeleteTaskCommand.java:
```diff
@@ -0,0 +1,58 @@
+            botStorage.deleteTask(taskNumber);
+            dateTable.deleteTaskOnDate(removeTask);
+            return ui.showDeleteTask(removeTask, taskList.getTotalTask());
+        }
```
Hi! I can suggest another way to improve readability in this case is to swap the if and else part
so you can do this: 
```
if (NumericChecker.isNumeric(description)) {
              //do something
} else {
              return exception.printNumericError("delete");
}
```

Perhaps you could look into my code and consider!

albertsutz commented on src/main/java/angela/command/GetTaskFromDateCommand.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Extracts events on a specific date command
+ */
+public class GetTaskFromDateCommand extends angela.command.Command {
```
I'm not sure whether I like the name or not. The name feels like a method rather than a class. Perhaps you could reconsider?

albertsutz replied:
This is also true for several of the commands classes.

albertsutz commented on src/main/java/angela/exception/BotException.java:
```diff
@@ -0,0 +1,46 @@
+     * Prints error when the bot command does not follow with description
+     *
+     * @param description The input command type
+     *
```
I don't think that line the blank space in line 12 is necessary.

albertsutz reviewed (commented):
Hi there, overall a really nice job! I can clearly see that you have put a lot of effort and the results really shows.

Perhaps some aspects that you might consider:
- Should the `/data/Duke.txt` be committed?
- You should add a blank line between the brief description and the parameters (in the javadocs). I see these in most of your documentation.

Good job and good luck for your next increments!

khoahre123 committed:
Fix code according to code review suggestion

khoahre123 committed:
Add assertion to all class

There are several methods that can result in unexpected behavior of the
code if someone change a variable or a block of code. These problems
mostly originated from import library

Hence, to aid future developer of the app, assertion is added so that
they won't accidently change anything that may cause harm in runtime

khoahre123 committed:
Merge pull request #1 from khoahre123/branch-A-Assertions

Add assertion to all class

khoahre123 committed:
Merge branch 'master' into branch-A-Assertions

khoahre123 committed:
Improve code quality

As the objective of the code till now is just simply to get the code to
work, the code is smell very bad. For new developer in the project, it
is impossible to know what happen in there.

Hence, I have:
* Extract method to avoid deep nesting
* Put number and string as static variable to increase understanding
* Clear the running path for easy reading

khoahre123 committed:
Merge pull request #2 from khoahre123/branch-A-CodeQuality

Improve code quality

khoahre123 committed:
Merge branch 'master' into branch-A-Assertions

khoahre123 committed:
Merge branch 'master' into branch-A-CodeQuality

khoahre123 committed:
Implement stream for sorting task

Currently, a for loop with if is used for sort task according to
requirement.

But as the number of tasks increase, a more effective way, and quicker
way is needed to improve our app efficiency

Therefore, stream is used as a replacement whenever there is a sorting
happen

Currently, it is done using sequential streaming, but a parallel one can
be used to increase speed if needed

khoahre123 committed:
Merge pull request #3 from khoahre123/branch-A-Streams

Implement stream for sorting task

khoahre123 committed:
Merge branch 'master' into branch-A-Assertions

khoahre123 committed:
Merge branch 'master' into branch-A-CodeQuality

khoahre123 committed:
Merge branch 'master' into branch-A-Streams

khoahre123 committed:
Create main.yml

khoahre123 committed:
Merge branch 'branch-A-Assertions'

khoahre123 committed:
Merge branch 'branch-A-CodeQuality'

khoahre123 committed:
Merge branch 'branch-A-Streams'

khoahre123 committed:
Update GUI when user start and terminate program

khoahre123 committed:
Add reminder extension for upcoming deadline

khoahre123 committed:
Merge pull request #5 from khoahre123/branch-B-Reminders

Add reminder extension for upcoming deadline

khoahre123 committed:
Merge branch 'master' into branch-B-Reminders

khoahre123 committed:
Merge branch 'branch-B-Reminders'

khoahre123 committed:
Update GUI of Angela

khoahre123 committed:
Finish GUI and chatbot-user interaction

khoahre123 committed:
Fix javadoc to follow coding standard

khoahre123 committed:
Add Ui picture

khoahre123 committed:
Edit user guide for release

khoahre123 committed:
Fix readme

khoahre123 committed:
Fix Ui.png

khoahre123 committed:
Fix gradle build

khoahre123 committed:
Fix assertion
